pkg/urbit/vere/io/ames.c

```diff
@@ -30,21 +33,57 @@
 */
   typedef struct _u3_ames {             //  packet network state
     u3_auto       car_u;                //  driver
+    u3_pier*      pir_u;                //  pier
     union {                             //
       uv_udp_t    wax_u;                //
       uv_handle_t had_u;                //
     };                                  //
-    c3_d          who_d[2];             //  identity
-    c3_o          fak_o;                //  fake keys
-    c3_s          por_s;                //  public IPv4 port
     c3_c*         dns_c;                //  domain XX multiple/fallback
     c3_d          dop_d;                //  drop count
     c3_d          fal_d;                //  crash count
     c3_w          imp_w[256];           //  imperial IPs
     time_t        imp_t[256];           //  imperial IP timestamps
     c3_o          imp_o[256];           //  imperial print status
+    c3_o          see_o;                //  can scry
+    c3_o          fit_o;                //  filtering active
+    c3_y          ver_y;                //  protocol version
+    c3_d          vet_d;                //  version mismatches filtered
+    c3_d          mut_d;                //  invalid mugs filtered
+    u3_panc*      pac_u;                //  packets pending forwards
+    c3_d          foq_d;                //  forward queue size
+    c3_d          fow_d;                //  forwarded count
+    c3_d          fod_d;                //  forwards dropped count
   } u3_ames;
 
+/* u3_head: ames packet header
+*/
+  typedef struct _u3_head {
+    c3_y ver_y;                         //  protocol version
+    c3_l mug_l;                         //  truncated mug hash of u3_body
+    c3_y sac_y;                         //  sender class
+    c3_y rac_y;                         //  receiver class
+    c3_o enc_o;                         //  encrypted?
+  } u3_head;
+
+/* u3_body: ames packet body
+*/
+  typedef struct _u3_body {
+    u3_noun sen;                        //  sender
+    u3_noun rec;                        //  receiver
+    u3_noun con;                        //  (jam [origin content])
```

Comment:
To the degree that it's practical, it'd be better to save c structures over nouns here. The sender and receiver could easily be `c3_d[2]`.

As structured, the content could be a pointer to a byte buffer, at the cost of some minor memory management responsibilities. But if the content is to be cue'd earlier in the process, it might not be practical to save byte buffer(s) (barring more flexible de/serialization routines than the current ones).